ink/ink/internal/types.go

```
@@ -0,0 +1,520 @@
+	switch {
+	case typeIs[TypeVar](ty1):
+		id := ty1.(TypeVar).id
+		ctx.substitutions[id] = ty2
+		fmt.Printf("$%d is resolved to %s\n", id, ty2.String(ctx))
+		return ty2
+	case typeIs[TypeVar](ty2):
+		id := ty2.(TypeVar).id
+		ctx.substitutions[id] = ty1
+		fmt.Printf("$%d is resolved to %s\n", id, ty1.String(ctx))
+		return ty1
+	case typeIs[TypeFunction](ty1) && typeIs[TypeFunction](ty1):
```

golangci-lint (1.64.8)

SA4000: identical expressions on the left and right side of the '&&' operator

(staticcheck)